tests/helpers/test_config_entry_oauth2_flow.py

```diff
@@ -963,3 +974,145 @@ async def test_oauth2_without_secret_init(
     client = await hass_client_no_auth()
     resp = await client.get("/auth/external/callback?code=abcd&state=qwer")
     assert resp.status == 400
+
+
+@pytest.mark.usefixtures("current_request_with_host")
+async def test_abort_if_oauth_with_pkce_rejected(
+    hass: HomeAssistant,
+    flow_handler: type[config_entry_oauth2_flow.AbstractOAuth2FlowHandler],
+    local_impl_pkce: config_entry_oauth2_flow.LocalOAuth2ImplementationWithPkce,
+    hass_client_no_auth: ClientSessionGenerator,
+) -> None:
+    """Check bad oauth token."""
+    flow_handler.async_register_implementation(hass, local_impl_pkce)
+    config_entry_oauth2_flow.async_register_implementation(
+        hass, TEST_DOMAIN, MockOAuth2Implementation()
+    )
+
+    result = await hass.config_entries.flow.async_init(
+        TEST_DOMAIN, context={"source": config_entries.SOURCE_USER}
+    )
+
+    assert result["type"] == data_entry_flow.FlowResultType.FORM
+    assert result["step_id"] == "pick_implementation"
+
+    # Pick implementation
+    result = await hass.config_entries.flow.async_configure(
+        result["flow_id"], user_input={"implementation": TEST_DOMAIN}
+    )
+
+    state = config_entry_oauth2_flow._encode_jwt(
+        hass,
+        {
+            "flow_id": result["flow_id"],
+            "redirect_uri": "https://example.com/auth/external/callback",
+        },
+    )
+
+    assert result["type"] == data_entry_flow.FlowResultType.EXTERNAL_STEP
+
+    assert result["url"].startswith(f"{AUTHORIZE_URL}?")
+    assert f"client_id={CLIENT_ID}" in result["url"]
+    assert "redirect_uri=https://example.com/auth/external/callback" in result["url"]
+    assert f"state={state}" in result["url"]
+    assert "scope=read+write" in result["url"]
+    assert "response_type=code" in result["url"]
+    assert "code_challenge=" in result["url"]
+    assert "code_challenge_method=S256" in result["url"]
+
+    client = await hass_client_no_auth()
+    resp = await client.get(
+        f"/auth/external/callback?error=access_denied&state={state}"
+    )
+    assert resp.status == 200
+    assert resp.headers["content-type"] == "text/html; charset=utf-8"
+
+    result = await hass.config_entries.flow.async_configure(result["flow_id"])
+
+    assert result["type"] == data_entry_flow.FlowResultType.ABORT
+    assert result["reason"] == "user_rejected_authorize"
+    assert result["description_placeholders"] == {"error": "access_denied"}
+
+
+# @pytest.mark.usefixtures("current_request_with_host")
+# async def test_if_oauth_with_pkce_adds_code_verifier_to_token_resolve(
+#     hass: HomeAssistant,
+#     flow_handler: type[config_entry_oauth2_flow.AbstractOAuth2FlowHandler],
+#     local_impl_pkce: config_entry_oauth2_flow.LocalOAuth2ImplementationWithPkce,
+#     hass_client_no_auth: ClientSessionGenerator,
+#     aioclient_mock: AiohttpClientMocker,
+# ) -> None:
+#     """Check pkce flow."""
+#     mock_platform(hass, f"{TEST_DOMAIN}.config_flow")
```

Comment:
Here's an example if you want to mock an integration with a config flow.

https://github.com/home-assistant/core/blob/eadff2938f7c8f1180f75ae4fcee470a9d812d33/tests/components/assist_satellite/conftest.py#L143-L164

https://github.com/home-assistant/core/blob/eadff2938f7c8f1180f75ae4fcee470a9d812d33/tests/components/assist_satellite/conftest.py#L170-L172

You probably need to adapt it depending on what we want to test.